README.md
# flip
Fork of flip v1.20 source to add support for compilation under MinGW


*flip* does newline conversions between **IX and MS-DOS conventions.  Uses a state
machine, so repeated conversions on the same file will do no harm.
Assumes the US ASCII character set in some places (search for 'ASCII').



TODO:
-----
- Make tests work under Windows


ChangeLog
2020-05-22  Jason Morgan <dev@cropwell.net>
	* Makefile: Add mingw target, add support for gcc under MinGW64,
	            Add install target for Windows
	* flip.h:   Add mingw defines, handle file renames under mingw
	
	* version 1.21

2008-07-26  James R. Van Zandt  <jrv@vanzandt.comcast.net>

	* Makefile (check): new target "check".  Include test-flip in
	distribution file.  Add comments giving command to build on a
	modern Unix system with a standard C compiler.

	* test-flip: new shell script to test flip

	* flip.h (SETFT): #include utime.h.  call utime with a struct
	utimbuf rather than an array.

	* flip.c: bring sccsid up to date.
	(dofile): handle case NEITHER. Transfer old file permissions and
	ownership to new file. Add return value to silence gcc warning.
	(mstoix): add parens to silence gcc warning

2008-07-25  James R. Van Zandt  <jrv@debian.org>

	* flip.h (PATHSIZE): set PATHSIZE from FILENAME_MAX (in stdio.h)
	if possible.
	
	* Makefile: version 1.20

2001-07-15  James R. Van Zandt  <jrv@debian.org>

	* flip.c: preserve file permissions.
	Accept input from stdin, write to stdout (thanks to "Dwayne
	C. Litzenberger" <dlitz@dlitz.net>) 

	* flip.1: document the above.

2001-07-13  James R. Van Zandt  <jrv@debian.org>

	* Makefile (DISTFILES): add "dist" target.
	* flip.c: help message inherits version number from Makefile
	* flip.c: use mkstemp instead of mktemp to reduce security risk.

	* version 1.19



flip.h
/* ::[[ @(#) flip.h 1.16 89/07/04 17:00:46 ]]:: */

/*
Copyright 1989 Rahul Dhesi, All rights reserved.

Checksum: 2652839101 (check or update with "brik")

ANSI-compatibility strategy:

1.   If ANSIPROTO is defined, function prototypes are used, ANSI-style.
     If it is not defined, regular function declarations (without a
     parameter list) are used.  This is achieved with the use of
     the PARMS macro.
2.   If STDINCLUDE is defined, ANSI-conformant header files are
     included.  Otherwise functions that would be declared in such
     header files are declared individually.
*/

#ifndef OK_NL

/* Define ANSIPROTO and/or STDINCLUDE here if needed */
#ifdef TURBOC
# define    ANSIPROTO
# define    STDINCLUDE
#endif /* TURBOC */



#include <stdio.h>

#ifndef PARMS
# ifdef ANSIPROTO
#  define   PARMS(x)    x
# else
#  define   PARMS(x)    ()
# endif
#endif

/************************************************************/
/*** change following definitions as needed for your system */
/************************************************************/
/*
BIGBUF      If defined, setvbuf() will be used to set input and
            output buffers to BIGBUF characters
BINCHAR     This must be a macro accepting a single character
            argument and returning nonzero if and only if that
            character is a binary character.  By default, BINCHAR
            is defined to look up a table, and the table is compiled
            in by defining USE_TABLE.
CHECK_BREAK If your operating system requires programs to check
            for user interrupts at intervals, define CHECK_BREAK to
            be a function or macro that checks for a user interrupt.
            Flip will call this at intervals.  If not needed,
            define CHECK_BREAK to be null.
DELFILE     Must be a function or macro that will accept one filenames
            and delete that file, returning zero on success, nonzero
            on failure.
GETFT       Must be a macro with usage GETFT(file,name,var) that will
            take an open file, a filename, and a variable name, and
            put in that variable the timestamp of the file.
INT_EXIT    If USE_SIG is defined, INT_EXIT must be defined to
            be the exit code returned by flip when its execution
            is aborted by a user interrupt.
MVFILE      Must be a function or macro that will take two filenames
            and rename the first file to the second file, returning
            zero on success, nonzero on failure.
NIX         If this symbol is defined, some things common to most
            implementations of **IX get defined and need not be
            individually defined.
PATHSIZE    size of buffer(s) to use to hold pathname(s)
PICKNAME    Do "#define PICKNAME" if you want the program to know its
            name and respond to it.  If PICKNAME is not defined, the
            program will assume that it is called "flip".
SETFT       Must be a function or macro with usage SETFT(file,name,var)
            that will take an open file, a filename, and a variable
            name, and set the timestamp of the file from the variable.
SIG_T       If USE_SIG is defined, then SIG_T must be defined
            (or typedef'd) to be the data type returned by
            (*signal).  Usually int or void.
SPEC_EXIT   Like SPEC_INIT, but is called just before flip exits.
SPEC_INIT   This is invoked to initialize flip.  Should be defined to
            be either a function call (with trailing semicolon) or
            null string.
SGFTIME     This must be a typedef for a data type suitable to store
            file times.
USE_TABLE   If this is defined, flip code will use a table look-up
            to decide which characters should be considered binary.
            See also BINCHAR.
ULTRIX_BUG  At least one version of Ultrix chokes on a trailing "b"
            in the mode string for fopen().  If ULTRIX_BUG is defined
            no trailing "b" will be used.
USE_SIG     If this is defined, then signal() will be used so flip
            may trap user interrupts and delete temporary files.
            If USE_SIG is defined, then SIG_T and INT_EXIT must also
            be defined (below).
WILDCARD    If this is defined, the function nextfile() will be called
            to expand wildcards.  If WILDCARD is not defined, no
            wildcard expansion will be done and nextfile() is not
            needed.  For a description of nextfile() see turboc.c.
*/

#ifdef FILENAME_MAX
#define PATHSIZE FILENAME_MAX
#endif /* FILENAME_MAX */

#ifdef TURBOC
#ifndef PATHSIZE
# define PATHSIZE    200
#endif /* PATHSIZE */
# define PICKNAME
# define BIGBUF      16384
# define DELFILE     remove
extern int MVFILE PARMS ((char *src, char *dest));            /* Turbo C */
# define USE_SIG
# define SIG_T void     /* return type from (*signal) */
# define INT_EXIT 127   /* status when a signal causes exit */
# define CHECK_BREAK  brktst();
void brktst PARMS ((void));
# define SPEC_INIT   spec_init();
# define SPEC_EXIT   spec_exit();
void spec_init PARMS ((void));
void spec_exit PARMS ((void));
# define WILDCARD

/* date and time */
# include <io.h>
  typedef struct ftime        SGFTIME;
# define GETFT(f,name,var)    getftime(fileno(f), &var)
# define SETFT(f,name,var)    setftime(fileno(f), &var)
#endif /* TURBOC */

#ifdef SYS_V
# define NIX         /* see below */
#ifndef PATHSIZE
# define PATHSIZE    200
#endif /* PATHSIZE */
# define BIGBUF      16384
# define  MVFILE(src,dest)    (!(!link(src,dest)&&!unlink(src)))
# define SIG_T int
#endif /* SYS_V */

#ifdef BSD
# define NIX         /* see below */
#ifndef PATHSIZE
# define PATHSIZE    1027
#endif /* PATHSIZE */
# undef  BIGBUF
# define  MVFILE     rename
# define SIG_T int
#endif /* BSD */

#ifdef MINGW
# define NIX         /* see below */
#ifndef PATHSIZE
# define PATHSIZE    1027
#endif /* PATHSIZE */
# undef  BIGBUF
# define  MVFILE(src, dest)     !((!DELFILE(dest) && !rename(src, dest)))
# define SIG_T int
#endif /* MINGW */

#ifdef VMS
#ifndef PATHSIZE
# define PATHSIZE    1024
#endif /* PATHSIZE */
# define PICKNAME
# define BIGBUF      16384
# define DELFILE     delete
# define  MVFILE     rename
# define USE_SIG
# define SIG_T int
# define INT_EXIT    127
# define SPEC_INIT
# define SPEC_EXIT
#endif /* VMS */

/************************************************************/
/*** remaining definitions should not need to be changed ****/
/************************************************************/

/* define ASCII character values for linefeed, carriage return, control Z */
#define  LF    10
#define  CR    13
#define  CTRLZ 26

/* error codes */
#define  ERRBINF  1  /* binary file */
#define  ERRSIG   2  /* interrupted by signal */

/* how to strip high bit */
#define  STRIP(c)    c &= 0x7f

/**** BINCHAR(c) returns nonzero if c is a binary char ****/
#ifndef BINCHAR
# define USE_TABLE      /* use internal table -- see flip.c */
# define BINCHAR(c)  bintab[c]
#endif

/**** mode strings for fopen() -- to work around Ultrix problem ****/
/**** (Thanks to Taso N. Devetzis for reporting this)           ****/
#ifdef ULTRIX_BUG
# define WB          "w"
# define RB          "r"
# define R_PL_B      "r+"
#else
# define WB          "wb"     /* write binary */
# define RB          "rb"     /* read binary */
# define R_PL_B      "r+b"    /* read + update binary */
#endif /* ULTRIX_BUG */

/* things common to most **IX systems */
#ifdef NIX
# define SPEC_INIT
# define SPEC_EXIT
# define CHECK_BREAK
# define PICKNAME
# define DELFILE     unlink
# define USE_SIG
# define INT_EXIT    127
# include <sys/types.h>          /* for file time */
# include <sys/stat.h>           /* for file time */
# include <utime.h>		 /* for utime() */
  typedef struct stat            SGFTIME;
# define GETFT(file,name,var)    fstat(fileno(file),&var)
# define SETFT(file,name,var)    do { struct utimbuf x; x.actime = var.st_atime; \
                                     x.modtime = var.st_mtime; utime (name, &x); \
                                  } while (0)
#endif /* NIX */

#endif /* OK_NL */


getopt.c
/* ::[[ @(#) getopt.c 1.5 89/07/02 00:18:19 ]]:: */
#ifndef LINT
static char sccsid[]="::[[ @(#) getopt.c 1.5 89/07/02 00:18:19 ]]::";
#endif

/*
Checksum:  505161377      (check or update this with "brik")
*/

/*
 * Here's something you've all been waiting for:  the AT&T public domain
 * source for getopt(3).  It is the code which was given out at the 1985
 * UNIFORUM conference in Dallas.  I obtained it by electronic mail
 * directly from AT&T.  The people there assure me that it is indeed
 * in the public domain.
 *
 * There is no manual page.  That is because the one they gave out at
 * UNIFORUM was slightly different from the current System V Release 2
 * manual page.  The difference apparently involved a note about the
 * famous rules 5 and 6, recommending using white space between an option
 * and its first argument, and not grouping options that have arguments.
 * Getopt itself is currently lenient about both of these things White
 * space is allowed, but not mandatory, and the last option in a group can
 * have an argument.  That particular version of the man page evidently
 * has no official existence, and my source at AT&T did not send a copy.
 * The current SVR2 man page reflects the actual behavor of this getopt.
 * However, I am not about to post a copy of anything licensed by AT&T.
 */

/*
Minor modifications by Rahul Dhesi 1989/03/06
*/

#include "flip.h"

#ifdef STDINCLUDE
# include <string.h>
#else
extern int strcmp PARMS ((char *, char *));
extern char *strchr PARMS ((char *, char));
#endif

/* Avoid possible compiler warning if we simply redefine NULL or EOF */
#define XNULL   0
#define XEOF (-1)

#define ERR(szz,czz) if(opterr){fprintf(stderr,"%s%s%c\n",argv[0],szz,czz);}

int   opterr = 1;
int   optind = 1;
int   optopt;
char  *optarg;

int
getopt(argc, argv, opts)
int   argc;
char  **argv, *opts;
{
   static int sp = 1;
   register int c;
   register char *cp;

   if(sp == 1)
      if(optind >= argc ||
         argv[optind][0] != '-' || argv[optind][1] == '\0')
         return(XEOF);
      else if(strcmp(argv[optind], "--") == XNULL) {
         optind++;
         return(XEOF);
      }
   optopt = c = argv[optind][sp];
   if(c == ':' || (cp=strchr(opts, c)) == XNULL) {
      ERR(": illegal option -- ", c);
      if(argv[optind][++sp] == '\0') {
         optind++;
         sp = 1;
      }
      return('?');
   }
   if(*++cp == ':') {
      if(argv[optind][sp+1] != '\0')
         optarg = &argv[optind++][sp+1];
      else if(++optind >= argc) {
         ERR(": option requires an argument -- ", c);
         sp = 1;
         return('?');
      } else
         optarg = argv[optind++];
      sp = 1;
   } else {
      if(argv[optind][++sp] == '\0') {
         sp = 1;
         optind++;
      }
      optarg = XNULL;
   }
   return(c);
}


flip.c
/* ::[[ @(#) flip.c 1.18 89/07/04 16:07:16 ]]:: */
#ifndef LINT
static char sccsid[]="::[[ @(#) flip.c 1.20 2008-07-26 ]]::";
#endif

/*
Copyright 1989 Rahul Dhesi, All rights reserved.

Checksum: 1217582374 (check or update with "brik")
*/

/*
Does newline conversions between **IX and MS-DOS conventions.  Uses a state
machine, so repeated conversions on the same file will do no harm.
Assumes the US ASCII character set in some places (search for 'ASCII').
*/

/* change contents of flip.h as needed */
#include "flip.h"

enum choices   { MSTOIX, IXTOMS, NEITHER };           /* conversion choices */
enum state     { NONE, SAWCR, SAWLF, SAWCRLF, SAWCTRLZ };

void usage PARMS ((void));
void give_help PARMS ((void));
void flip_exit PARMS ((int));
int getopt PARMS ((int argc, char **argv, char *options));
void doarg PARMS ((char *, enum choices));
int dofile PARMS ((char *, enum choices));
char *nextfile PARMS ((int, char *, int));
int ixtoms PARMS ((FILE *, FILE *));
int mstoix PARMS ((FILE *, FILE *));
void error PARMS ((char *, char *));
void setup_sigs PARMS ((void));
void cleanup PARMS ((int));
int mkstemp PARMS ((char *));

#ifdef STDINCLUDE
# include <stdlib.h>
# include <string.h>
#else
void *malloc PARMS ((int));
void exit PARMS ((int));
char *strcpy PARMS ((char *, char *));
#endif

#ifdef IX
#include <sys/types.h>
#include <sys/stat.h>		/* stat and chmod */
#include <unistd.h>
#endif

#ifdef NDEBUG
# define assert(c)
#else
# ifdef STDINCLUDE
#  include <assert.h>
# else
#  define assert(c)  if(!(c)) \
                  fprintf(stderr,"assert error %s:%d\n",__FILE__,__LINE__)
# endif /* STDINCLUDE */
#endif /* NDEBUG */

#ifdef USE_TABLE
char *bintab;
#endif

#ifdef USE_SIG
int got_sig;               /* will indicate if signal received */
#endif

char *myname = NULL;

int exitstat = 0;          /* exit status */
int verbose = 0;           /* set by -v option */
int touch = 0;             /* set by -t option */
int strip = 0;             /* set by -s option */
int bintoo = 0;            /* set by -b option */
int ztrunc = 0;            /* set by -z option */
int use_stdio = 0;	   /* set by "-" filename */

int main (argc, argv)
int argc;
char **argv;

{
   int option;
   extern int optind;
   int i;
   enum choices which = NEITHER;
#ifdef USE_TABLE
#define TABSIZ    256
   char table[TABSIZ];
#endif

#ifdef PICKNAME
   register char *p; /* temp pointer for finding our name */
   register char *arg0 = *argv;
#endif /* PICKNAME */

   SPEC_INIT            /* optional initialization */

#ifdef USE_SIG
   setup_sigs();
#endif

#ifdef PICKNAME
# define STRCMP(a,op,b)    (strcmp(a,b) op 0)
   p = arg0 + strlen(arg0);

   if (p != arg0) {     /* if program name is defined */
      while (p != arg0 && *p != '/' && *p != '\\')
         p--;
      assert ((p - arg0) <= strlen (arg0));
      if (p != arg0)
         p++;

      /* p now points to trailing name component, or nothing */
      myname = p;
      while (*p != '\0' && *p != '.') {   /* ASCII convert to lowercase */
         if (*p >= 'A' && *p <= 'Z') {
            *p = (*p - 'A' + 'a');
         }
         p++;
      }
      if (p != myname && *p == '.')
         *p = '\0';     /* remove trailing .exe or .com under MS-DOS etc. */

      if (STRCMP(myname,==,"toix"))
         which = MSTOIX;
      else if (STRCMP(myname,==,"toms"))
         which = IXTOMS;
   } else
   myname = "flip";
#else
   myname = "flip";
#endif /* PICKNAME */

argv[0] = myname;                /* for use by getopt */

#ifdef USE_TABLE
/* table to find out which characters are binary */
   for (i = 0;  i < TABSIZ;  i++) {
      if ( (i < 7) || (i > 13 && i < 26) || (i > 126)) /*ASCII binary chars*/
         table[i] = 1;
      else
         table[i] = 0;
   }
   bintab = table;
#endif /* USE_TABLE */

   if (argc < 2) {
      usage();
      flip_exit (1);
   }

   while ((option = getopt (argc, argv, "umhvtsbz")) != EOF) {
      switch (option) {
       case 'u': which = MSTOIX;  break;
       case 'm': which = IXTOMS; break;
       case 'h': give_help(); flip_exit (0);
       case 'v': verbose = 1;  break;
       case 't': touch = 1;  break;
       case 's': strip = 1;  break;
       case 'b': bintoo = 1;  break;
       case 'z': ztrunc = 1;  break;
       default:  usage(); flip_exit (1);
      }
   }

   switch (which) {
    case MSTOIX:
      /* printf ("converting to **ix format\n"); */
      break;
    case IXTOMS:
      /* printf ("converting to msdos format\n"); */
      break;
    default:
      fprintf (stderr, "%s: error: -u or -m is required\n", myname);
      flip_exit (1);
      break;
   }

   if (argc <= optind) {
      fprintf (stderr, "%s: error: filenames are needed\n", myname);
      flip_exit (1);
   }

   for (i = optind;  i < argc;  i++)
      doarg (argv[i], which);

   return (exitstat);
}


/*
Does conversion for one argument, calling dofile with wildcards
expanded.  Updates exitstat in case of error.
*/

void doarg (arg, which)
char *arg;
enum choices which;
{
#ifdef WILDCARD
   char *this_file;

   nextfile (0, arg, 0);
   while ((this_file = nextfile(1, (char *) NULL, 0)) != NULL) {
      exitstat |= dofile (this_file, which);
   }
#else
   exitstat |= dofile (arg, which);
#endif /* WILDCARD */
}

#ifdef USE_SIG
# include <signal.h>
#endif

FILE *outfile;       /* make it visible to both dofile and cleanup */

/*
Here we have filename and an option.  We call a different routine
for each type of conversion.  This way more types of conversions
can be easily added in the future.
*/

int dofile (fname, which)
char *fname;
enum choices which;
{
   FILE *infile;
   char tfname[PATHSIZE];
   SGFTIME timestamp;  /* save file timestamp here, restore later */
   int errstat = 0;
   char *p;             /* temp file ptr */
#ifdef IX
   struct stat ifilestat, ofilestat;
#endif

#ifdef USE_SIG
   if (got_sig)
      flip_exit (INT_EXIT);
#endif

   if(STRCMP(fname,==,"-"))
      use_stdio = 1;
   if(!use_stdio) {
      /* if writable, open for reading */
      if ((infile = fopen (fname, R_PL_B)) != NULL) {
         fclose (infile);
         infile = fopen (fname, RB);
#ifdef IX
	 if (stat (fname, &ifilestat)) {
	   /* can't get the file's permissions */
	   char *buf = malloc(strlen(fname)+strlen(myname)+3);
	   if (buf) {
	     sprintf(buf, "%s: %s", myname, fname);
	     perror(buf);
	   }
	   ifilestat.st_mode = 0644;
	   ifilestat.st_uid = geteuid();
	   ifilestat.st_gid = getegid();
	 }
#endif
      }
   } else {
      infile = stdin;
   }
	   
   if (infile == NULL) {
      error (fname, ": can't open");
      return (1);
   }
   if(use_stdio) {
     outfile = stdout;
   } else {
     /*
       to make temp file in same dir as original, we make p point to
       the filename component of fname, put '\0' there, and strcat the
       temp name to it */
     strcpy (tfname, fname);
     p = tfname + strlen(tfname);
     
     while (p != tfname && *p != '/' && *p != '\\')
       p--;
     if (p != tfname)
       p++;
     *p = '\0';

#define  TEMPLATE    "XXXXXX"
#ifdef IX
     strcat (tfname, TEMPLATE);

     {
       int fd = mkstemp(tfname);
       if (fd == -1 || (outfile = fdopen(fd, WB)) == NULL) {
	 fclose (infile);
	 error (fname, ": skipped, could not open temporary file");
	 return (1);
       }
     }
#else
     {
       char template[7];
       strcpy (template, TEMPLATE);
       strcat (tfname, mktemp (template));
     }

     outfile = fopen (tfname, WB);
     if (outfile == NULL) {
       fclose (infile);
       error (fname, ": skipped, could not open temporary file");
       return (1);
     }
#endif
   }

   if (!touch)
      GETFT (infile, fname, timestamp);      /* save current timestamp */

   assert (which == IXTOMS || which == MSTOIX);

#ifdef BIGBUF
   setvbuf (infile,  (char *) NULL, _IOFBF, BIGBUF);
   setvbuf (outfile, (char *) NULL, _IOFBF, BIGBUF);
#endif /* BIGBUF */

   switch (which) {
    case IXTOMS:
      errstat = ixtoms (infile, outfile);
      break;
    case MSTOIX:
      errstat = mstoix (infile, outfile);
      break;
    case NEITHER:
      ;
   }

   fclose (infile);

   switch (errstat) {
    case ERRBINF:
      fclose (outfile);
      DELFILE (tfname);
      fprintf (stderr, "%s: binary file, not converted\n", fname);
      return (1);
      /* break; */  /* unreachable code */
#ifdef USE_SIG
    case ERRSIG:
      fclose (outfile);
      DELFILE (tfname);
      flip_exit (INT_EXIT);
      break;
#endif
    default:
      ;
   }

   assert (errstat == 0);

   if (!ferror(outfile) && fflush(outfile) != EOF && fclose(outfile) != EOF) {
      int moved;
#ifdef IX
      if (stat (tfname, &ofilestat)) {
	/* can't get the file's permissions */
	char *buf = malloc(strlen(tfname)+strlen(myname)+3);
	if (buf) {
	  sprintf(buf, "%s: %s", myname, tfname);
	  perror(buf);
	}
      }
      if (ifilestat.st_gid != ofilestat.st_gid ||
	  ifilestat.st_uid != ofilestat.st_uid)
	if (chown(tfname, ifilestat.st_uid, ifilestat.st_gid)) {
	  /* can't set the file's ownership */
	  char *buf = malloc(strlen(tfname)+strlen(myname)+3);
	  if (buf) {
	    sprintf(buf, "%s: %s", myname, tfname);
	    perror(buf);
	  }
	}
      if (ifilestat.st_mode != ofilestat.st_mode)
	if (chmod(tfname, ifilestat.st_mode)) {
	  /* can't set the file's permissions */
	  char *buf = malloc(strlen(tfname)+strlen(myname)+3);
	  if (buf) {
	    sprintf(buf, "%s: %s", myname, tfname);
	    perror(buf);
	  }
	}
#endif /* IX */
      //      DELFILE (fname);
      if (!use_stdio) {
         moved = MVFILE (tfname, fname);
         if (moved == 0) {
            FILE *fptr;
            if (!touch && (fptr = fopen (fname, RB)) != NULL) {
               SETFT (fptr, fname, timestamp);
               fclose (fptr);
            }
            if (verbose) {
               printf ("%s\n", fname);
               fflush (stdout);
            }
            return (0);
         } else {
            error (fname, ": not converted, could not rename temp file");
            DELFILE (tfname);
            return (1);
         }
      }
   } else {
      fclose (outfile); /* outfile was not closed, so close it here */
      return (1);
   }
   return (1);			/* silences gcc warning */
}

/* convert from ms-dos to **ix format */
int mstoix (infile, outfile)
FILE *infile;           /* input file   */
FILE *outfile;          /* output file  */
{
   int c;
   enum state state = NONE;

   /* lone LF => unchanged, lone CR => unchanged,
      CR LF => LF, ^Z at end means EOF; ^Z elsewhere => unchanged */

   while (1) {       /* break out on EOF only */
      while ((c = getc (infile)) != EOF) {
#ifdef USE_SIG
         if (got_sig)
            return (ERRSIG);
#endif
         if (!bintoo && BINCHAR(c))
            return (ERRBINF);
         if (strip)
            STRIP(c);
         switch (c) {
          case LF:
            CHECK_BREAK
            putc (c, outfile); if (state == SAWCR) state = NONE; break;
          case CR:
            state = SAWCR; break;
          case CTRLZ:
            if (state == SAWCR) putc (CR, outfile);
            state = SAWCTRLZ; goto saweof;
          default:
            if (state == SAWCR) { state = NONE; putc (CR, outfile); }
            putc (c, outfile);
            break;
         }
      }
 saweof:
      /* exit outer loop only on EOF or ^Z as last char */
      if (
          (ztrunc && (state == SAWCTRLZ))
          || (c = getc (infile)) == EOF
         )
         break;
      else
         ungetc (c, infile);
      if (state == SAWCTRLZ)
         putc (CTRLZ, outfile);
   }
   return (0);
}

/* convert from **ix to ms-dos format */
int ixtoms (infile, outfile)
FILE *infile;           /* input file   */
FILE *outfile;          /* output file  */
{
   int c;
   enum state state = NONE;

   /* LF => CR LF, but leave CR LF alone */
   while ((c = getc (infile)) != EOF) {
#ifdef USE_SIG
      if (got_sig)
         return (ERRSIG);
#endif
      if (!bintoo && BINCHAR(c))
         return (ERRBINF);
      if (strip)
         STRIP(c);
      switch (c) {
       case LF:
         CHECK_BREAK
         if (state == SAWCR)
            state = NONE;
         else
            putc (CR, outfile);
         putc (LF, outfile);
         break;
       case CR:
         state = SAWCR; putc (c, outfile); break;
       case CTRLZ:
         if (ztrunc)
            return (0);
         /* FALL THROUGH */
       default:
         state = NONE; putc (c, outfile); break;
      }
   }
   return (0);
}

/* set up signal handler for selected signals */
#ifdef USE_SIG
void setup_sigs ()
{
# ifdef SIGPIPE
   if (signal (SIGPIPE, SIG_IGN) != SIG_IGN)
      signal (SIGPIPE, cleanup);
# endif
# ifdef SIGHUP
   if (signal (SIGHUP, SIG_IGN) != SIG_IGN)
      signal (SIGHUP, cleanup);
# endif
# ifdef SIGQUIT
   if (signal (SIGQUIT, SIG_IGN) != SIG_IGN)
      signal (SIGQUIT, cleanup);
# endif
# ifdef SIGINT
   if (signal (SIGINT, SIG_IGN) != SIG_IGN)
      signal (SIGINT, cleanup);
# endif
# ifdef SIGTERM
   if (signal (SIGTERM, SIG_IGN) != SIG_IGN)
      signal (SIGTERM, cleanup);
# endif
}

/* set flag on signal */
void cleanup(sig)
int sig;
{
   signal (sig, SIG_IGN);     /* needed for flaky System V Release 2 */
   got_sig = 1;
   signal (sig, cleanup);     /* ditto */
}
#endif /* USE_SIG */

#define  ERRSIZE     200

/* prints error message via perror */
void error (msg1, msg2)
char *msg1, *msg2;
{
   char buf[ERRSIZE];
   strcpy (buf, myname);
   strcat (buf, ": ");
   strcat (buf, msg1);
   strcat (buf, msg2);
   perror (buf);
   fflush (stderr);
}

/* gives brief usage message */
void usage()
{
   fprintf (stderr,
"usage:  %s [-u | -m] [other options] file ...  (or \"%s -h\" for more help)\n",
 myname, myname);
}

/* gives help screen */

void give_help()
{
printf ("\
File interchange program flip version " VERSION ".  Copyright 1989 Rahul Dhesi,\n\
All rights reserved.  Both noncommercial and commercial copying, use, and\n\
creation of derivative works are permitted in accordance with the\n\
requirements of the GNU license.  This program does newline conversions.\n");

printf ("\n\
   Usage:     flip -umhvtsbz file ...\n\
\n\
One of -u, -m, or -h is required;  others are optional.  See user manual.\n");

printf ("\n\
   -u     convert to **IX format (CR LF => LF, lone CR or LF unchanged,\n\
          trailing control Z removed, embedded control Z unchanged)\n\
   -m     convert to MS-DOS format (lone LF => CR LF, lone CR unchanged)\n\
   -h     give this help message\n\
   -v     be verbose, print filenames as they are processed\n\
   -t     touch files (don't preserve timestamps)\n\
   -s     strip high bit\n\
   -b     convert binary files too (else binary files are left unchanged)\n\
   -z     truncate file at first control Z encountered\n\
");
#ifdef PICKNAME
printf ("\n\
May be invoked as \"toix\" (same as \"flip -u\") or \"toms\" (same as \"flip -m\").\n\
");
#endif
return;
}

/* normal exits go through here -- atexit would be nice but not portable */
void flip_exit(stat)
int stat;
{
   SPEC_EXIT
   exit (stat);
}

/*
special code for **IX -- not in use because can't properly handle
SIGINT while /bin/mv is executing
*/

#ifdef NIX
# ifndef MVFILE
int MVFILE (src, dest)
char *dest;
char *src;
{
   char cmd[2 * PATHSIZE];
   sprintf (cmd, "/bin/mv %s %s", src, dest);
   return (system(cmd));
}
# endif /* MVFILE */
#endif /* NIX */
